synapse/app/homeserver.py

```diff
@@ -241,6 +242,22 @@ def start_listening(self):
             )
             logger.info("Metrics now running on 127.0.0.1 port %d", config.metrics_port)
 
+    @defer.inlineCallbacks
+    def post_startup_check(self):
+        all_users_native = yield self.get_datastore().are_all_users_on_domain(
+            self.hostname
+        )
+        if not all_users_native:
+            sys.stderr.write(
+                "\n"
+                "******************************************************\n"
+                "Found users in database not native to %s!\n"
+                "You cannot changed a synapse server_name after it's been configured\n"
+                "******************************************************\n"
+                "\n"
```

Comment:
Should this not also log to the log? I don't think we want to assume that stderr goes anywhere sensible once we've daemonized.